pkg/kademlia/kademlia.go

```diff
@@ -239,26 +239,30 @@ func (k *Kad) Start(ctx context.Context) error {
 
 func (k *Kad) connectBootnodes(ctx context.Context) {
 	var count int
-
 	for _, addr := range k.bootnodes {
-		if count == 3 {
+		if count >= 3 {
 			return
 		}
 
 		if _, err := p2p.Discover(ctx, addr, func(addr ma.Multiaddr) (stop bool, err error) {
 			k.logger.Tracef("connecting to bootnode %s", addr)
-			_, err = k.p2p.ConnectNotify(ctx, addr)
+			bzzAddress, err := k.p2p.Connect(ctx, addr)
 			if err != nil {
 				if !errors.Is(err, p2p.ErrAlreadyConnected) {
 					k.logger.Debugf("connect fail %s: %v", addr, err)
 					k.logger.Warningf("connect to bootnode %s", addr)
+					return false, err
 				}
 				return false, nil
 			}
+
+			if err := k.connected(ctx, bzzAddress.Overlay); err != nil {
```

Comment:
this has the unwanted side effect that you would try to announce a bootnode. presumably unnecessary

Reply:
I noticed this but did not want to change the existing logic. The bootnodes used to be connected with connectNotify which triggers connected handler, which I believe also announces. `k.connect` & `k.Connected` both announces. I really think it is a bad approach to have to think about these different cases all the time. It is very easy to make a mistake and really hard to see the implications of adding some code to kademlia service, for example.
So, if you are sure that announce is not required, I would erase it gladly. But then I need to copy some code again, wihich is different only in announce.

Reply:
as we discussed before, different connection flows have different requirements, so as much as I'm a big fan of refactoring and reusing of repeated code, sometimes it is inevitable.